### PR TITLE
Layout array (based on telescope_layout)

### PR DESCRIPTION
**Pull request by RaulRPrado**
I adapted the telescope_layout code into the simtools library.
We have now simtools.layout that contains layout_array and telescope_data.

The idea is that LayoutArray will be used inside ArrayModel or CorsikaConfig to deal with the telescope positions.

I added an application make_regular_layouts as an example of using layout_array and also to create the ECSV files for the regular arrays, needed for, e.g, trigger rate simulations.

By now, the ECSV files containing the array data will be at the data/layout directory. In the future this data should be put somewhere else, preferably in a DB.

I worked here under the assumption that there are few options for array layout, which is what we expect to happen in the end.
By now one can easily build regular layout (1SST, 4SST, 1 MST, ...) or Prod5 layoutfor South and North. Playing around with 
different layout (for optimization studies, for example) is possible but it shouldn't work out-of-the-box.   

The simplest way the layout_array can be used is using the classmethod functions that builds a pre-defined layout from its name. Example: 

layout = LayoutArray.fromLayoutArrayName('South-Prod5')

or 

layout = LayoutArray.fromLayoutArrayName('North-4MST')

After that, one can do layout.getCorsikaInputList to get the piece of text that goes into the CORSIKA input file.

### CODE REVIEW

**Comment by orelgueta** [comment]
Notice we also have a plotting application for the arrays in the "site" branch, we can include it once this PR is merged.

**Comment by RaulRPrado** [comment]
> Notice we also have a plotting application for the arrays in the "site" branch, we can include it once this PR is merged.

Good to know, thanks. I won't write any array plotting tools then.

**Comment by RaulRPrado** [comment]
Site parameters are stored in a yaml file in data/layout/site_parameters.yml. However, this is only needed to build new layout array, apart from the pre-defined ones, which will be used much more often.

These data could go into a DB in the future.

**Comment by orelgueta** [comment]
The site parameters are already in the DB. We have a collection called "site" and a function in db_handler called "getSiteParameters".

**Comment by RaulRPrado** [comment]
> The site parameters are already in the DB. We have a collection called "site" and a function in db_handler called "getSiteParameters".

Which ones? Sorry, I used misleading terms.
What I meant here are location of the center of the array. Like:

    epsg: 32719
    centerLongitude: !astropy.units.Quantity
        value: -70.316345
        unit: !astropy.units.Unit {unit: deg}
    centerLatitude: !astropy.units.Quantity
        value: -24.683429
        unit: !astropy.units.Unit {unit: deg}
    centerAltitude: !astropy.units.Quantity
        value: 2162.35
        unit: !astropy.units.Unit {unit: m}
    centerNorthing: !astropy.units.Quantity
        value: 7269466.0
        unit: !astropy.units.Unit {unit: m}
    centerEasting: !astropy.units.Quantity
        value: 366822.0
        unit: !astropy.units.Unit {unit: m}
    corsikaObsLevel: !astropy.units.Quantity
        value: 2147
        unit: !astropy.units.Unit {unit: m}
    corsikaSphereRadius:
        LST: !astropy.units.Quantity
            value: 12.5
            unit: !astropy.units.Unit {unit: m}
        MST: !astropy.units.Quantity
            value: 9.6
            unit: !astropy.units.Unit {unit: m}
        SST: !astropy.units.Quantity
            value: 3.0
            unit: !astropy.units.Unit {unit: m}
    corsikaSphereCenter:
        LST: !astropy.units.Quantity
            value: 16
            unit: !astropy.units.Unit {unit: m}
        MST: !astropy.units.Quantity
            value: 9
            unit: !astropy.units.Unit {unit: m}
        SST: !astropy.units.Quantity
            value: 3.25
            unit: !astropy.units.Unit {unit: m}

**Comment by orelgueta** [comment]
Yes, the centre of each array is in the DB. The CORSIKA spheres aren't, but I can add them quickly if you wish.

**Comment by GernotMaier** [comment]
Note that the spheres are not really a  site parameter. They have been added to allow to calculate the full corsika config from the telescope list. In the case here, I guess that we can get the sphere from the model db.

Do we need this actually as a model parameter or could we calculate it dynamically from the size of the primary mirror?

**Comment by RaulRPrado** [comment]
Ok, so should we fix this before merging or do we make a new PR later?

It's easy for me to grab the site parameters from the DB, but I'd prefer to have the sphere radius from there too.

File: applications/make_regular_arrays.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        with open(siteParsFile, 'r') as f:
+            sitePars = yaml.load(f)
+    except FileNotFoundError:
+        msg = 'Site parameter file ({}) could not be opened/read'.format(siteParsFile)
+        logger.error(msg)
+        raise FileNotFoundError(msg)
+
+    # Telescope distances for 4 tel square arrays
+    # !HARDCODED
+    telescopeDistance = {'LST': 57.5 * u.m, 'MST': 70 * u.m, 'SST': 80 * u.m}
```
Comment (GernotMaier) [inline applications/make_regular_arrays.py:88]: @RaulRPrado Why is this needed? Is this for tests? So you define telescope distances for square layouts here. The distances are very small.

Comment (RaulRPrado) [reply applications/make_regular_arrays.py:88]: These are the square arrays that we use for trigger rate simulations.
It's very specific and it will only be used in this application. And still, it's an application that is used only once.

Comment (GernotMaier) [reply applications/make_regular_arrays.py:88]: OK - and we only do single-telescope trigger there, no array trigger, correct? I am just surprised by the small telescope distances.

Comment (RaulRPrado) [reply applications/make_regular_arrays.py:88]: For LST, the standard is stereo trigger with square array. For MST and SST we use single telescope.
But these are the standards, which doesn't mean we can't have the option of building square array, for whatever reason we need.
I will double check the distances.

**Comment by GernotMaier** [comment]
Seen this before:?

```
python applications/make_regular_arrays.py
ERROR::config(l83)::get::Config does not contain dataLocation
Traceback (most recent call last):
  File "/workdir/external/gammasim-tools/applications/make_regular_arrays.py", line 75, in <module>
    siteParsFile = io.getDataFile('layout', 'site_parameters.yml')
  File "/workdir/external/gammasim-tools/simtools/io_handler.py", line 162, in getDataFile
    return Path(cfg.get('dataLocation')).joinpath(parentDir).joinpath(fileName)
  File "/workdir/external/gammasim-tools/simtools/config.py", line 84, in get
    raise KeyError()
KeyError
```

**Comment by RaulRPrado** [comment]
> Seen this before:?
> 
> ```
> python applications/make_regular_arrays.py
> ERROR::config(l83)::get::Config does not contain dataLocation
> Traceback (most recent call last):
>   File "/workdir/external/gammasim-tools/applications/make_regular_arrays.py", line 75, in <module>
>     siteParsFile = io.getDataFile('layout', 'site_parameters.yml')
>   File "/workdir/external/gammasim-tools/simtools/io_handler.py", line 162, in getDataFile
>     return Path(cfg.get('dataLocation')).joinpath(parentDir).joinpath(fileName)
>   File "/workdir/external/gammasim-tools/simtools/config.py", line 84, in get
>     raise KeyError()
> KeyError
> ```

The config structure was slightly changed. The template should be updated by your local version is still wrong.
Just change testDataLocation -> dataLocation.

File: applications/make_regular_arrays.py (hunk 1 above)
Comment (RaulRPrado) [reply applications/make_regular_arrays.py:88]: In sim_telarray there are CORSIKA input files for square arrays for all telescope tyles.
You can find, for example, INPUTS_CTA-trg-test-prod3-MST4_proton. 
The distances were taken from there. They might be outdated since they are from prod3.

**Comment by GernotMaier** [comment]
> > Seen this before:?
> > ```
> > python applications/make_regular_arrays.py
> > ERROR::config(l83)::get::Config does not contain dataLocation
> > Traceback (most recent call last):
> >   File "/workdir/external/gammasim-tools/applications/make_regular_arrays.py", line 75, in <module>
> >     siteParsFile = io.getDataFile('layout', 'site_parameters.yml')
> >   File "/workdir/external/gammasim-tools/simtools/io_handler.py", line 162, in getDataFile
> >     return Path(cfg.get('dataLocation')).joinpath(parentDir).joinpath(fileName)
> >   File "/workdir/external/gammasim-tools/simtools/config.py", line 84, in get
> >     raise KeyError()
> > KeyError
> > ```
> 
> The config structure was slightly changed. The template should be updated by your local version is still wrong.
> Just change testDataLocation -> dataLocation.

Ok - config.yml is in .gitignore, that's why any changes are not synced. Is that intended?

File: applications/make_regular_arrays.py (hunk 1 above)
Comment (GernotMaier) [reply applications/make_regular_arrays.py:88]: Couldn't we have that as a configuration in the DB - trigger array layouts?

**Comment by GernotMaier** [comment]
`python applications/make_regular_arrays.py` runs now without any issues.

Is the metadata correct? What is `#     unit: *id002`?

telescope_positions-South-1SST.ecsv:
```
# %ECSV 0.9
# ---
# datatype:
# - {name: telescope_name, datatype: string}
# - {name: pos_x, unit: m, datatype: float64}
# - {name: pos_y, unit: m, datatype: float64}
# - {name: pos_z, unit: m, datatype: float64}
# - {name: lat, unit: deg, datatype: float64}
# - {name: lon, unit: deg, datatype: float64}
# - {name: utm_east, unit: deg, datatype: float64}
# - {name: utm_north, unit: deg, datatype: float64}
# - {name: alt, unit: m, datatype: float64}
# meta:
#   EPSG: 32719
#   center_alt: !astropy.units.Quantity
#     unit: &id001 !astropy.units.Unit {unit: m}
#     value: 2162.35
#   center_easting: !astropy.units.Quantity
#     unit: *id001
#     value: 366822.0
#   center_lat: !astropy.units.Quantity
#     unit: &id002 !astropy.units.Unit {unit: deg}
#     value: -24.683429
#   center_lon: !astropy.units.Quantity
#     unit: *id002
#     value: -70.316345
#   center_northing: !astropy.units.Quantity
#     unit: *id001
#     value: 7269466.0
#   corsika_obs_level: !astropy.units.Quantity
#     unit: *id001
#     value: 2147.0
#   corsika_sphere_center:
#     LST: !astropy.units.Quantity
#       unit: *id001
#       value: 16.0
#     MST: !astropy.units.Quantity
#       unit: *id001
#       value: 9.0
#     SST: !astropy.units.Quantity
#       unit: *id001
#       value: 3.25
#   corsika_sphere_radius:
#     LST: !astropy.units.Quantity
#       unit: *id001
#       value: 12.5
#     MST: !astropy.units.Quantity
#       unit: *id001
#       value: 9.6
#     SST: !astropy.units.Quantity
#       unit: *id001
#       value: 3.0
# schema: astropy-2.0
telescope_name pos_x pos_y pos_z lat lon utm_east utm_north alt
S-01 0.0 0.0 0.0 -24.683429 -70.316345 366821.99919380073 7269466.016937867 2143.75
```

**Comment by RaulRPrado** [comment]
> > > Seen this before:?
> > > ```
> > > python applications/make_regular_arrays.py
> > > ERROR::config(l83)::get::Config does not contain dataLocation
> > > Traceback (most recent call last):
> > >   File "/workdir/external/gammasim-tools/applications/make_regular_arrays.py", line 75, in <module>
> > >     siteParsFile = io.getDataFile('layout', 'site_parameters.yml')
> > >   File "/workdir/external/gammasim-tools/simtools/io_handler.py", line 162, in getDataFile
> > >     return Path(cfg.get('dataLocation')).joinpath(parentDir).joinpath(fileName)
> > >   File "/workdir/external/gammasim-tools/simtools/config.py", line 84, in get
> > >     raise KeyError()
> > > KeyError
> > > ```
> > 
> > 
> > The config structure was slightly changed. The template should be updated by your local version is still wrong.
> > Just change testDataLocation -> dataLocation.
> 
> Ok - config.yml is in .gitignore, that's why any changes are not synced. Is that intended?

Yes. config.yml is not in the repository, it is a local, user prepared file.
In the repo we have a config_template.yml only.

Same thing with dbDetails.yml

**Comment by GernotMaier** [comment]
config_template.yml requires then also the change from testDataLocation to dataLocation?

**Comment by RaulRPrado** [comment]
> Is the metadata correct? What is `# unit: *id002`?
> 

Yes, this is how astropy saves the units. I tested reading these files and it works fine.
Note that u.m and u.deg is written only once, and them astropy uses this id for the following appearences.

It's more important for these files to be consistent with astropy than humanly readable.

**Comment by RaulRPrado** [comment]
> config_template.yml requires then also the change from testDataLocation to dataLocation?

Yes, because I mistakenly didn't do it. I just pushed a correction, sorry about that.

File: applications/make_regular_arrays.py (hunk 1 above)
Comment (orelgueta) [reply applications/make_regular_arrays.py:88]: Yes, I think we can add it into the sites collection.

**Comment by orelgueta** [comment]
> Note that the spheres are not really a site parameter. They have been added to allow to calculate the full corsika config from the telescope list. In the case here, I guess that we can get the sphere from the model db. Do we need this actually as a model parameter or could we calculate it dynamically from the size of the primary mirror?

Yes, I think we could calculate it from the primary mirror, but it won't be a completely trivial calculation (we do not have a parameter for the size of the primary mirror). Not sure what's best, setting it as a parameter that maybe is useful to tune (doubt it, but OK) or just calculate a reasonable value.

**Comment by GernotMaier** [comment]
> > Note that the spheres are not really a site parameter. They have been added to allow to calculate the full corsika config from the telescope list. In the case here, I guess that we can get the sphere from the model db. Do we need this actually as a model parameter or could we calculate it dynamically from the size of the primary mirror?
> 
> Yes, I think we could calculate it from the primary mirror, but it won't be a completely trivial calculation (we do not have a parameter for the size of the primary mirror). Not sure what's best, setting it as a parameter that maybe is useful to tune (doubt it, but OK) or just calculate a reasonable value.

Let's keep it simple and put it as a parameter in the DB - these are N values for N telescope types. I am opening an issue implement some on-the-fly calculation of R.

**Comment by GernotMaier** [comment]
> > Is the metadata correct? What is `# unit: *id002`?
> 
> Yes, this is how astropy saves the units. I tested reading these files and it works fine.
> Note that u.m and u.deg is written only once, and them astropy uses this id for the following appearences.
> 
> It's more important for these files to be consistent with astropy than humanly readable.

Good! It is is the first time I see it; I guess it is way to keep it more effective. OK!

**Comment by GernotMaier** [comment]
What is the decision now on the data files? I would advocate to keep them outside of the code and in the DB.

Note that:
- the prod5 files are outdated as we have some telescope added (North)
- there is a group working on the coordinate system definition for CTAO; so it might not be WGS84 and we 'master' list will be hedged by CTAO (there is a first draft: https://gitlab.cta-observatory.org/cta-science/cta-science/-/tree/master -- but obviously we don't want to have markdown files without any metadata)

I am otherwise happy to get this PR merged.

**Comment by RaulRPrado** [comment]
> What is the decision now on the data files? I would advocate to keep them outside of the code and in the DB.
> 
> Note that:
> 
> * the prod5 files are outdated as we have some telescope added (North)
> * there is a group working on the coordinate system definition for CTAO; so it might not be WGS84 and we 'master' list will be hedged by CTAO (there is a first draft: https://gitlab.cta-observatory.org/cta-science/cta-science/-/tree/master -- but obviously we don't want to have markdown files without any metadata)
> 
> I am otherwise happy to get this PR merged.

I'm working now on collecting the site parameters from the DB. A few of them will still be hardcoded, until Orel manage to add them to the DB, probably next week.

I don't think having the up-to-date files or the right coordinate system at this point is a priority. It's more important to have these features implemented. And then we can update whatever we need, whenever the final decisions are made.

I may be done by the end of the day and then we can merge the PR.

**Comment by RaulRPrado** [comment]
The site_pars are being read from the DB now. We don't use a site_parameters.yml file anymore.
In the applications, you can find a dict with the hardcoded pars that have to go into the DB (probably in a future PR).

@orelgueta, please merge it if you don't find any problem.
